### ISSUE TITLE
support interrupt nesting

### ISSUE THREAD

**Issue by japaric** [japaric/msp430-rtfm#1]
Currently interrupt nesting is disabled so a higher priority interrupt can't preempt a lower priority one. At some point we should provide the option to enable nesting but probably in an opt-in manner so that people that don't want to use it, to e.g. keep binary size low, won't have to "pay" for it.

I think the tricky part of supporting this is that interrupt are not configurable but instead fixed and device specific. The problem is that the `app!` macro expects the user to input the priority to be able to compute the ceilings but that doesn't make much sense in this and it would even be error prone: the user may input a priority that doesn't match the actual value and that would led to unsafety.

**Comment by pftbest** [comment]
We already have all the priority numbers, it is the same as interrupt Nr.

**Comment by pftbest** [comment]
Or i'm missing the point completely, need to read more about "ceilings".
The doc says:

> 2.2.3.3 Interrupt Nesting
 Interrupt nesting is enabled if the GIE bit is set inside an interrupt service routine. When interrupt nesting is enabled, any interrupt occurring during an interrupt service routine will interrupt the routine, regardless of the interrupt priorities.

**Comment by japaric** [comment]
> We already have all the priority numbers, it is the same as interrupt Nr.

Yes, but the `app!` macro needs to now the values at *expansion time* because it decides *types* (`&mut T` vs `Resource<T>`) based on the value of priorities. Ceilings are easy to compute (they are just the biggest value across the priorities of the tasks that access a resource), but their exact value is not too important here because MSP430 devices don't have something equivalent to BASEPRI which effectively lets you temporarily raise the priority of an interrupt to prevent preemption from *some* interrupts.

That being say we could compromise here and use `Resource` everywhere but that would require using `claim` which is not too ergonomic because it requires a closure: `let x = *r.STATE;` vs `let x = r.STATE.claim(t, |state, _| *state);`
